codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -66,412 +70,1049 @@
-                    ReferenceCountUtil.release(frame);
-                    throw new IllegalArgumentException("Stream id must not be set on the frame. Was: "
-                        + frame.streamId());
-                }
-                if (!isStreamIdValid(streamId())) {
-                    if (!(frame instanceof Http2HeadersFrame)) {
-                        ReferenceCountUtil.release(frame);
-                        throw new IllegalArgumentException("The first frame must be a headers frame. Was: "
-                            + frame.name());
-                    }
-                    frame = new ChannelCarryingHeadersFrame((Http2HeadersFrame) frame, this);
-                    // Handle errors on stream creation
-                    promise.addListener(this);
-                } else {
-                    frame.streamId(streamId());
-                }
-                writeFromStreamChannel(frame, promise, false);
-            } else if (msg instanceof Http2GoAwayFrame) {
-                ChannelPromise promise = ctx.newPromise();
-                promise.addListener(this);
-                writeFromStreamChannel(msg, promise, false);
-            } else {
-                ReferenceCountUtil.release(msg);
-                throw new IllegalArgumentException("Message must be an Http2GoAwayFrame or Http2StreamFrame: " + msg);
-            }
+        @Override
+        public Channel parent() {
+            return ctx.channel();
         }
 
         @Override
-        protected void doWriteComplete() {
-            flushFromStreamChannel();
+        public boolean isRegistered() {
+            return registered;
         }
 
         @Override
-        protected EventExecutor preferredEventExecutor() {
-            return ctx.executor();
+        public SocketAddress localAddress() {
+            return parent().localAddress();
         }
 
         @Override
-        protected void bytesConsumed(final int bytes) {
-            ctx.write(new DefaultHttp2WindowUpdateFrame(bytes).streamId(streamId()));
+        public SocketAddress remoteAddress() {
+            return parent().remoteAddress();
         }
 
         @Override
-        public void operationComplete(ChannelFuture future) throws Exception {
-            Throwable cause = future.cause();
-            if (cause != null) {
-                pipeline().fireExceptionCaught(cause);
-                close();
-            }
+        public ChannelFuture closeFuture() {
+            return closePromise;
         }
-    }
 
-    /**
-     * Wraps the first {@link Http2HeadersFrame} of local/outbound stream. This allows us to get to the child channel
-     * when receiving the {@link Http2StreamActiveEvent} from the frame codec. See {@link #onStreamActive}.
-     */
-    private static final class ChannelCarryingHeadersFrame implements Http2HeadersFrame {
+        @Override
+        public long bytesBeforeUnwritable() {
+            // TODO: Do a proper impl
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)